### PR TITLE
xds: Server implementation

### PR DESCRIPTION
**Pull request by easwars**
- Contains the initial implementation which defines the server side xDS API.
- Does not contain much xDS specific functionality. This will be added in subsequent PRs.
- Also contains an e2e test.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc-go/3999)

### CODE REVIEW

**Comment by easwars** [comment]
@dfawley : I'm assigning only @menghanl because I have sent you too many PRs of late :). But would appreciate your comments as well, if you have any. Thanks.

**Review by menghanl — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc-go/3999)* status: 2 of 3 files reviewed, 4 unresolved discussions (waiting on @dfawley, @easwars, and @menghanl)

___
*[xds/server.go, line 195 at r1](https://reviewable.io/reviews/grpc/grpc-go/3999#-MKukrRr-oGHcGWc6WtV:-MKukrRs-3BlkiMWf7VY:b3uqhxc) ([raw file](https://github.com/grpc/grpc-go/blob/39970cb14660d5380350bca84e2c5b3a9e0d7142/xds/server.go#L195)):*
> ```Go
> 	}
> 
> 	lis, err := net.Listen(opts.Network, opts.Address)
> ```

Move `net.Listen`, `WatchLDS` into a `newListenerWrapper()` function?
___
*[xds/server.go, line 271 at r1](https://reviewable.io/reviews/grpc/grpc-go/3999#-MKumi_98JKFzkd4Ethw:-MKumi_98JKFzkd4Ethx:b-hsxj9q) ([raw file](https://github.com/grpc/grpc-go/blob/39970cb14660d5380350bca84e2c5b3a9e0d7142/xds/server.go#L271)):*
> ```Go
> 	s.quit.Fire()
> 	s.gs.GracefulStop()
> 	s.xdsC.Close()
> ```

`if s.xdsC != nil`
___
*[xds/server.go, line 297 at r1](https://reviewable.io/reviews/grpc/grpc-go/3999#-MKumswQ39lwxlq03-cO:-MKumswQ39lwxlq03-cP:b-gxfjya) ([raw file](https://github.com/grpc/grpc-go/blob/39970cb14660d5380350bca84e2c5b3a9e0d7142/xds/server.go#L297)):*
> ```Go
> 	cancelWatch func()
> 
> 	// Mutex to protect the certificate providers.
> ```

Delete those and other provider related code?

It helps keep this change small, and also the next PR will have all the security code.
___
*[xds/server_test.go, line 145 at r1](https://reviewable.io/reviews/grpc/grpc-go/3999#-MKunMFJ2FKai8OtdUoo:-MKunMFJ2FKai8OtdUop:bo2y1jh) ([raw file](https://github.com/grpc/grpc-go/blob/39970cb14660d5380350bca84e2c5b3a9e0d7142/xds/server_test.go#L145)):*
> ```Go
> 	// a single ServerOption in the test, we expect to see 3 in the call to
> 	// grpc.NewGRPCServer().
> 	wantServerOpts := 3
> ```

```go
wantServerOpts = len(serverOpts) + 2
```

**Review by easwars — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc-go/3999)* status: 2 of 3 files reviewed, 4 unresolved discussions (waiting on @dfawley and @menghanl)

___
*[xds/server.go, line 195 at r1](https://reviewable.io/reviews/grpc/grpc-go/3999#-MKukrRr-oGHcGWc6WtV:-MKvO29G30qREmCaGEtI:b4u2qk7) ([raw file](https://github.com/grpc/grpc-go/blob/39970cb14660d5380350bca84e2c5b3a9e0d7142/xds/server.go#L195)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Move `net.Listen`, `WatchLDS` into a `newListenerWrapper()` function?
</blockquote></details>

I could possible move the `net.Listen` into a `newListenerWrapper()` function, but that doesn't seem to provide much value. We would a new function which does the same thing and would have to return an error. And here we would have to call the new function and check for the returned err value and return for non-nil errors.

As for moving the call to `WatchListener` into the function, we would have to pass a few things: `xdsClient`, `logger`,  `goodUpdate` etc. The only thing really expected of the `listenerWrapper` it to cancel the watch when it is being closed.  If you think it would be better, I can rename the `cancelWatch` field in `listenerWrapper` to something more generic, like `closeHook` so indicate that this is just something that the `listenerWrapper` to invoke when it is closed.
___
*[xds/server.go, line 271 at r1](https://reviewable.io/reviews/grpc/grpc-go/3999#-MKumi_98JKFzkd4Ethw:-MKvH064CWw24M95uyz9:b-896fix) ([raw file](https://github.com/grpc/grpc-go/blob/39970cb14660d5380350bca84e2c5b3a9e0d7142/xds/server.go#L271)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

`if s.xdsC != nil`
</blockquote></details>

Done.
___
*[xds/server.go, line 297 at r1](https://reviewable.io/reviews/grpc/grpc-go/3999#-MKumswQ39lwxlq03-cO:-MKvGzYR16XDvjGC82k9:b-896fix) ([raw file](https://github.com/grpc/grpc-go/blob/39970cb14660d5380350bca84e2c5b3a9e0d7142/xds/server.go#L297)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Delete those and other provider related code?

It helps keep this change small, and also the next PR will have all the security code.
</blockquote></details>

Done.
___
*[xds/server_test.go, line 145 at r1](https://reviewable.io/reviews/grpc/grpc-go/3999#-MKunMFJ2FKai8OtdUoo:-MKvHFP4A5Zd2_VTBqEU:b-896fix) ([raw file](https://github.com/grpc/grpc-go/blob/39970cb14660d5380350bca84e2c5b3a9e0d7142/xds/server_test.go#L145)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

```go
wantServerOpts = len(serverOpts) + 2
```

</blockquote></details>

Done.

**Comment by easwars** [comment]
Ping ...

**Review by menghanl — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc-go/3999)* status: all files reviewed, 1 unresolved discussion (waiting on @dfawley and @easwars)

___
*[xds/server.go, line 195 at r1](https://reviewable.io/reviews/grpc/grpc-go/3999#-MKukrRr-oGHcGWc6WtV:-MLFHCF33a_tOdPybxR4:b5ogspl) ([raw file](https://github.com/grpc/grpc-go/blob/39970cb14660d5380350bca84e2c5b3a9e0d7142/xds/server.go#L195)):*
<details><summary><i>Previously, easwars (Easwar Swaminathan) wrote…</i></summary><blockquote>

I could possible move the `net.Listen` into a `newListenerWrapper()` function, but that doesn't seem to provide much value. We would a new function which does the same thing and would have to return an error. And here we would have to call the new function and check for the returned err value and return for non-nil errors.

As for moving the call to `WatchListener` into the function, we would have to pass a few things: `xdsClient`, `logger`,  `goodUpdate` etc. The only thing really expected of the `listenerWrapper` it to cancel the watch when it is being closed.  If you think it would be better, I can rename the `cancelWatch` field in `listenerWrapper` to something more generic, like `closeHook` so indicate that this is just something that the `listenerWrapper` to invoke when it is closed.
</blockquote></details>

I was just thinking that this function `Serve` doesn't need to know any of that (WatchListener() and wait for one response).
It just need to wait for the listener wrapper to be initialized and ready for use. Then call grpcserver.Serve on the listener wrapper.
If we make `newListenerWrapper()` a blocking call, this `Serve` will look cleaner.

And `newListenerWrapper()` would need to do more things later (like handling security config).
Potentially we may want to move it to its own file (maybe?)

But I don't feel strongly about this now. It's quite simple. Maybe you have plans to handle the updates differently.
___
*[xds/server.go, line 211 at r2](https://reviewable.io/reviews/grpc/grpc-go/3999#-MLFF_fU0Q4LiTu4VPWd:-MLFF_fU0Q4LiTu4VPWe:bl80i2p) ([raw file](https://github.com/grpc/grpc-go/blob/180d99734a0074386496c9f7001a67500e198857/xds/server.go#L211)):*
> ```Go
> 	// address as the resource name.
> 	// TODO(easwars): Check if literal IPv6 addresses need an enclosing [].
> 	name := fmt.Sprintf("grpc/server?udpa.resource.listening_address=%s", opts.Address)
> ```

Make a const for this string, and explain the format a bit.

**Review by easwars — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc-go/3999)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @dfawley and @menghanl)

___
*[xds/server.go, line 195 at r1](https://reviewable.io/reviews/grpc/grpc-go/3999#-MKukrRr-oGHcGWc6WtV:-MLFVtA16N35LveHAoQj:b-896fix) ([raw file](https://github.com/grpc/grpc-go/blob/39970cb14660d5380350bca84e2c5b3a9e0d7142/xds/server.go#L195)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

I was just thinking that this function `Serve` doesn't need to know any of that (WatchListener() and wait for one response).
It just need to wait for the listener wrapper to be initialized and ready for use. Then call grpcserver.Serve on the listener wrapper.
If we make `newListenerWrapper()` a blocking call, this `Serve` will look cleaner.

And `newListenerWrapper()` would need to do more things later (like handling security config).
Potentially we may want to move it to its own file (maybe?)

But I don't feel strongly about this now. It's quite simple. Maybe you have plans to handle the updates differently.
</blockquote></details>

Done.
___
*[xds/server.go, line 211 at r2](https://reviewable.io/reviews/grpc/grpc-go/3999#-MLFF_fU0Q4LiTu4VPWd:-MLFVuXEDIGvbcU5QutU:b-896fix) ([raw file](https://github.com/grpc/grpc-go/blob/180d99734a0074386496c9f7001a67500e198857/xds/server.go#L211)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Make a const for this string, and explain the format a bit.
</blockquote></details>

Done.

**Comment by easwars** [comment]
Thanks for the review @menghanl.

@dfawley : Please let me know if you want to take a look or if can I go ahead and merge. Thanks.

File: xds/server.go
Comment (dfawley) [inline xds/server.go:86]: Is this necessary?  Won't "tcp" always work?

File: xds/server.go
Comment (dfawley) [inline xds/server.go:129]: Is this information (and below) about the connection to the management server important to a user of the package?

File: xds/server.go
Comment (dfawley) [inline xds/server.go:133]: This comment seems more relevant on the type than the constructor.

File: xds/server.go
Comment (dfawley) [inline xds/server.go:126]: `s/a gRPC server/an xds-enabled gRPC server/`?

Maybe remove the bits about the interceptors as those are implementation details?

**Review by dfawley — COMMENTED** [review]
Looks good,just a few nits.

File: xds/server.go
Comment (easwars) [reply xds/server.go:19]: Done.

**Comment by easwars** [comment]
Finally, GA is green.

File: xds/server.go
Comment (dfawley) [reply xds/server.go:86]: Sure, let's go with that (hardcode to `"tcp"` for now).

File: xds/xds.go
Comment (dfawley) [inline xds/xds.go:26]: Will we also need a server example?  (Maybe later?)

File: xds/internal/test/xds_server_integration_test.go
Comment (dfawley) [inline xds/internal/test/xds_server_integration_test.go]: Maybe make this happen in the goroutine?  Just `t.Errorf` this and remove `errCh`?

**Comment by easwars** [comment]
Oops... I thought you were Ok with the changes. So, once I fixed the minor nits, I merged it.
I will address these comments in a follow-up PR.
I will anyways be adding the security integration to the server.

File: xds/xds.go
Comment (easwars) [reply xds/xds.go:26]: Sometimes github beats me totally. I don't understand why I can't reply to the above comment.

> Sure, let's go with that (hardcode to "tcp" for now).
Will do.

> Will we also need a server example? (Maybe later?)
Currently since the server doesn't have much functionality, I think we can wait till the other security integration work is complete and then we will have an example.

File: xds/internal/test/xds_server_integration_test.go
Comment (easwars) [reply xds/internal/test/xds_server_integration_test.go]: So, I initially had the `Serve` in a separate goroutine and the other stuff in the main test goroutine. And for some reason that just wasn't working on GA. The `Serve` goroutine never got scheduled and the `Dial` never succeeded  (because the server actually starts listening on that port only after `Serve` receives a Listener response).

Comment (dfawley) [reply xds/internal/test/xds_server_integration_test.go]: This doesn't make sense to me.  `Dial` should hit many synchronization points, so the `Serve` goroutine should get scheduled even if there's only 1 core.  I think this should be understood further.  Can you reproduce this failure mode locally with `-cpu 1`?

**Pull request merged**
